vsintegration/src/FSharp.Editor/Formatting/IndentationService.fs

```diff
@@ -12,35 +12,89 @@ open Microsoft.CodeAnalysis.Formatting
-                        | ' ' -> loop (column + 1) (spaces + 1)
-                        | '\t' -> loop (column + 1) (((spaces / tabSize) + 1) * tabSize)
-                        | _ -> spaces
-                Some (loop 0 0)
+
+        let rec tryFindLastNoneEmptyToken (line: TextLine) = maybe {
+           let! options = optionsOpt
+           let defines = CompilerEnvironment.GetCompilationDefinesForEditing(filePath, options.OtherOptions |> Seq.toList)
+           let tokens = Tokenizer.tokenizeLine(documentId, sourceText, line.Start, filePath, defines)
+
+           let rec loop (tokens: FSharpTokenInfo list) =
+               match tokens with
+               | [] -> None
+               | x::xs ->
+                   if x.Tag = FSharpTokenTag.WHITESPACE then
+                       loop xs
+                   else Some x
+
+           return! loop (List.rev tokens)
+        }
+
+        let (|Eq|_|) y x =
+            if x = y then Some()
+            else None
+
+        let (|NeedIndent|_|) (token: FSharpTokenInfo) =
+            match token.Tag with
+            | Eq FSharpTokenTag.EQUALS
+            | Eq FSharpTokenTag.LARROW
+            | Eq FSharpTokenTag.RARROW
+            | Eq FSharpTokenTag.LPAREN
+            | Eq FSharpTokenTag.LBRACK
+            | Eq FSharpTokenTag.LBRACK_BAR
+            | Eq FSharpTokenTag.LBRACK_LESS
+            | Eq FSharpTokenTag.LBRACE
+            | Eq FSharpTokenTag.BEGIN
+            | Eq FSharpTokenTag.DO
+            | Eq FSharpTokenTag.FUNCTION
+            | Eq FSharpTokenTag.THEN
+            | Eq FSharpTokenTag.ELSE
+            | Eq FSharpTokenTag.STRUCT
+            | Eq FSharpTokenTag.CLASS
+            | Eq FSharpTokenTag.TRY -> Some ()
+            | _ -> None
+
+        maybe {
+            // No indentation on the first line of a document
+            if lineNumber = 0 then return! None
+            else
+                match tryFindPreviousNonEmptyLine lineNumber with
+                | None -> return 0
+                | Some previousLine ->
+                    let rec loop column spaces =
+                        if previousLine.Start + column >= previousLine.End then
+                            spaces
+                        else
+                            match previousLine.Text.[previousLine.Start + column] with
+                            | ' ' -> loop (column + 1) (spaces + 1)
+                            | '\t' -> loop (column + 1) (((spaces / tabSize) + 1) * tabSize)
+                            | _ -> spaces
+
+                    let lastIndent = loop 0 0
+
+                    let! lastToken = tryFindLastNoneEmptyToken previousLine
+                    return
+                        match lastToken with
+                        | NeedIndent -> (lastIndent/tabSize + 1) * tabSize
+                        | _ -> lastIndent
+        }
 
     interface ISynchronousIndentationService with
```

Comment:
I don't know the different but seems like C# and VB also use `ISynchronousIndentationService`